LFS-254: As a user, I can see how my query matched each resource in the quick search results

IntegralProgrammer opened the pull request:
Enhances the *quick search bar* by displaying the context of the match in the suggestion UI with the matching content in **bold typeface**.

IntegralProgrammer committed:
Implements LFS-254: As a user, I can see how my query matched each resource in the quick search results

OrdiNeu commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -79,10 +80,45 @@ function SearchBar(props) {
       .catch(handleError);
   }
 
+  let addMatchingContent = (query, json) => {
```
This should not be handled in the frontend -- as it is, it's sending off one request per row in a synchronous, blocking manner, and obtaining a MB of data in total (depending on what's being returned). Instead, see if you can offload this logic to `QueryHandler.java`, which is in the backend. Since the `quickSearch` function there is specifically for this SearchBar, it can be altered as needed.

IntegralProgrammer replied:
This functionality has been moved to the back-end by commit `31ec9b7` :heavy_check_mark:

OrdiNeu commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -241,6 +311,14 @@ function SearchBar(props) {
                             </Typography>
                             {getElementName(element)}
                           </div>)}
+                          secondary={(<div>
```
A `<div>` element provided to the `secondary` prop of a ListItemText will cause the warning:
`Warning: validateDOMNesting(...): <div> cannot appear as a descendant of <p>.`.
There are two possible fixes:
1. Use the `disableTypography` option
```suggestion
                          disableTypography
                          secondary={(<div>
```
2. Switch from using `<div>` to `<React.Fragment>`
Either one will work.

IntegralProgrammer replied:
Implemented this change in commit `2a86d54` by switching from `<div>` to `<React.Fragment>` :heavy_check_mark:

OrdiNeu commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -241,6 +311,14 @@ function SearchBar(props) {
                             {getElementName(element)}
                           </div>)}
+                          secondary={(<div>
+                            {getElementQueryMatchKey(element)}
```
If you move the `if ("queryMatch" in element)` logic here, you can remove the redundancy of the check in this and the following three functions by calling them with `getElementQueryMatchKey(element["queryMatch"])`.

IntegralProgrammer replied:
Implemented by commits `4d5deda`, `380842f` :heavy_check_mark:

OrdiNeu reviewed (changes requested):
This works, but sends a lot of data&requests to&from the server -- see if you can move the suggestion logic to the backend.

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -79,10 +80,45 @@ function SearchBar(props) {
+          if ('value' in data[key]) {
+            let match_status = false;
+            if (data[key]['value'].toString().toLowerCase().indexOf(query.toLowerCase()) != -1) {
+              let match_index = data[key]['value'].toString().toLowerCase().indexOf(query.toLowerCase());
```
To reduce the number of ifs:
```
if (data[key]?.['value']?.toString().toLowerCase().indexOf(query.toLowerCase()) != -1) {
```
(please see https://github.com/tc39/proposal-optional-chaining via @OrdiNeu )

IntegralProgrammer replied:
Implemented by commit `6bce36b`. Please note that `!=` operator needed to be changed to a `>` operator as `undefined != -1` will evaluate to `true` while `undefined > -1` will evaluate to `false`.

marta- reviewed (changes requested):
Works well according to the currents specs, but the current specs need some improvements:
* When the user query matches the resource more than once, always show the first match (e.g. according to the ordering of the questions in the form)
* Highlight the matched string with a background (choose a light yellow from `theme.pallette`)

IntegralProgrammer committed:
Replace <div> with <React.Fragment> to eliminate validateDOMNesting warning

IntegralProgrammer committed:
Update modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx

Switch from `<b>` to `<span>` for emphasizing result text that matches search query.

Co-Authored-By: Marta Girdea <marta.girdea@gmail.com>

IntegralProgrammer committed:
Implement highlightedText class

IntegralProgrammer committed:
Removed code redundancy by moving ("queryMatch" in element) into a single ternary operator as opposed to four different IF statements

IntegralProgrammer committed:
Remove redundancy by only referencing element["queryMatch"] once

IntegralProgrammer commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -241,6 +311,14 @@ function SearchBar(props) {
                             </Typography>
                             {getElementName(element)}
                           </div>)}
+                          secondary={(<div>
```
Implemented this change in commit `2a86d54` by switching from `<div>` to `<React.Fragment>` :heavy_check_mark:

IntegralProgrammer committed:
Reduce number of nested IF statements by using Optional Chaining

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -79,10 +80,40 @@ function SearchBar(props) {
+    let processNext = (data) => {
+      for (let key in data) {
+        if (data[key]?.['value']?.toString().toLowerCase().indexOf(query.toLowerCase()) > -1) {
+          let match_index = data[key]['value'].toString().toLowerCase().indexOf(query.toLowerCase());
```
This is too long to be repeated. I suggest:
```
let match_index = data[key]?.[value]....
if (match_index > -1) {
...
```

IntegralProgrammer replied:
Good idea! Implemented in commit `c507243` :heavy_check_mark:

IntegralProgrammer committed:
Removed repeated query match logic - code simplification

IntegralProgrammer committed:
Add break statement to stop matching after the first match within a form. This way, when a query matches a resource more than once, the first match (according to the ordering of the questions in the form) is the one that is shown. This assumes that the order of elements in .deep.json is identical to that of the displayed form.

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -79,10 +80,41 @@ function SearchBar(props) {
+            'matchText' : data[key]['value'].toString().slice(match_index, match_index + query.length),
+            'matchAfter' : data[key]['value'].toString().slice(match_index + query.length)
+          };
+          break;
```
`Object.values(data)[0]` retrieves the first value in `data` (equivalent to `data[key]` for the first `key`) or `undefined` if `data` is empty. This is easier to read than a `for` loop with a `break` after the first run. It's not supported in  IE, but I think our code already has JS code too advanced for IE anyway (right @sdumitriu @OrdiNeu ?).

IntegralProgrammer committed:
Highlight matching text with theme.palette.warning.light color

IntegralProgrammer commented:
> Works well according to the currents specs, but the current specs need some improvements:
> 
>     * When the user query matches the resource more than once, always show the first match (e.g. according to the ordering of the questions in the form)

Addressed by commit `1e543fe`

> 
>     * Highlight the matched string with a background (choose a light yellow from `theme.pallette`)

Addressed by commit `a3cc44a`

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/headerStyle.jsx:
```diff
@@ -122,6 +122,10 @@ const headerStyle = theme => ({
+  },
+  highlightedText: {
+    fontWeight: "bold",
+    backgroundColor: theme.palette.warning.light
```
```suggestion
    backgroundColor: theme.palette.warning.light,
    padding: "1px 2px",
    borderRadius: "4px"
```
(just trying to make it look a bit smoother smoother)

IntegralProgrammer committed:
Update modules/homepage/src/main/frontend/src/themePage/Navbars/headerStyle.jsx

Highlighting enhancements

Co-Authored-By: Marta Girdea <marta.girdea@gmail.com>

IntegralProgrammer committed:
Moved lookup of matching data for a query string (ie. patient id, family name, etc), and subsequent query result data augmentation, from the front-end to the back-end

sashaandjic commented:
ready for testing?

marta- commented:
@sashaandjic Yes, but the code is not in the main repo, so you can't create a VM in our testing environment.

sashaandjic commented:
> @sashaandjic Yes, but the code is not in the main repo, so you can't create a VM in our testing environment.

Ah that's why I couldn't find it.

IntegralProgrammer commented:
> > @sashaandjic Yes, but the code is not in the main repo, so you can't create a VM in our testing environment.
> 
> Ah that's why I couldn't find it.

@sashaandjic Sorry about that, I copied everything to the `LFS-254` branch under the `ccmbioinfo/lfs` repository so you should now be able to test.

sashaandjic reviewed (commented):
the vm is running on http://172.20.4.126:8080/
When I enter "44" in the search bar I get this:
![image](https://user-images.githubusercontent.com/23243244/74562471-e9f2fc80-4f38-11ea-99a0-a66886607707.png)
When I type "440" I get the correct answer:
![image](https://user-images.githubusercontent.com/23243244/74562528-03944400-4f39-11ea-96a4-5e6e009f29a5.png)

IntegralProgrammer commented:
> the vm is running on http://172.20.4.126:8080/
> When I enter "44" in the search bar I get this:
> ![image](https://user-images.githubusercontent.com/23243244/74562471-e9f2fc80-4f38-11ea-99a0-a66886607707.png)
> When I type "440" I get the correct answer:
> ![image](https://user-images.githubusercontent.com/23243244/74562528-03944400-4f39-11ea-96a4-5e6e009f29a5.png)

I tested it in Firefox and I am getting similar behavior but a slightly different error message

![LFS254_fail44](https://user-images.githubusercontent.com/5349728/74565809-0b0b1b80-4f40-11ea-94b0-4bad6ad22c00.png)

![LFS254_working440](https://user-images.githubusercontent.com/5349728/74565827-18280a80-4f40-11ea-8c60-2f9544852d12.png)

Does this problem happen right from the beginning (after the VM is first created and started) or does it take time before this problem first happens?

sashaandjic commented:
@IntegralProgrammer  the problem happens right form the beginning

IntegralProgrammer committed:
Commit to address searchbar bug throwing error: java.lang.Boolean cannot be cast to java.lang.String

IntegralProgrammer commented:
@sashaandjic Please re-test and let me know if there are more bugs. This bug should be fixed by commit `623aa91`

sashaandjic commented:
tested on http://172.20.4.126:8080/

I found one problem that is also present in dev branch http://172.20.4.63:8080/ where the search does not bring Chemotherapy data when I search for 144 or 317517 
![image](https://user-images.githubusercontent.com/23243244/74867319-754a0480-5322-11ea-976d-60c27e93eba9.png). This will be logged as a separate bug.

marta- commented:
@sashaandjic The issue is known and in progress: 
https://phenotips.atlassian.net/browse/LFS-282

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -147,6 +148,32 @@ function SearchBar(props) {
+      return element[3].slice(0, MAX_CONTEXT_MATCH) + "...";
+    }
+  }
+
```
This logic (adjusting the context length and adding `...`) should also move to the backend.

Once we make Lucene highlighting work properly, we're going to get context information that already looks how we want it. Let's be diligent and remove this post-processioning from JS to make the transition smoother.

To clarify, `element[1]` and `element[3]` should already arrive from the backend at max 8 characters long and with `...` where applicable.

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -77,21 +168,28 @@ public void init(Bindings bindings)
+                augmentedResults = augmentIterator(results, true, this.urlDecode(quickQuery));
             } else {
                 results = Collections.emptyIterator();
+                augmentedResults = augmentIterator(results, false, "");
```
Just a note: the duplicated calls should be cleaned up as part of https://phenotips.atlassian.net/browse/LFS-246 .

sashaandjic commented:
tested

marta- commented:
Closing in favour of #168 .
